skada/deep/_graph_alignment.py

```diff
@@ -0,0 +1,201 @@
+# Author: Theo Gnassounou <theo.gnassounou@inria.fr>
+#
+# License: BSD 3-Clause
+
+import torch
+
+from skada.deep.base import (
+    BaseDALoss,
+    DomainAwareCriterion,
+    DomainAwareModule,
+    DomainAwareNet,
+    DomainBalancedDataLoader,
+)
+from skada.deep.callbacks import ComputeMemoryBank, MemoryBankInit
+from skada.deep.losses import gda_loss, nap_loss
+
+from .modules import DomainClassifier
+
+
+class SPALoss(BaseDALoss):
+    """Loss SPA.
+
+    This loss tries to minimize the divergence between features with
+    adversarial method. The weights are updated to make harder
+    to classify domains (i.e., remove domain-specific features).
+
+    See [36]_ for details.
+
+    Parameters
+    ----------
+    target_criterion : torch criterion (class), default=None
+        The initialized criterion (loss) used to compute the
+        adversarial loss. If None, a BCELoss is used.
+    reg_adv : float, default=1
+        Regularization parameter for adversarial loss.
+    reg_gsa : float, default=1
+        Regularization parameter for graph alignment loss
+    reg_nap : float, default=1
+        Regularization parameter for nap loss
+
+    References
+    ----------
+    .. [36] Xiao et. al. SPA: A Graph Spectral Alignment Perspective for
+            Domain Adaptation. In Neurips, 2023.
+    """
+
+    def __init__(
+        self,
+        domain_criterion=None,
+        memory_features=None,
+        memory_outputs=None,
+        K=5,
+        reg_adv=1,
+        reg_gsa=1,
+        reg_nap=1,
+    ):
+        super().__init__()
+        if domain_criterion is None:
+            self.domain_criterion_ = torch.nn.BCELoss()
+        else:
+            self.domain_criterion_ = domain_criterion
+
+        self.reg_adv = reg_adv
+        self.reg_gsa = reg_gsa
+        self.reg_nap = reg_nap
+        self.K = K
+        self.memory_features = memory_features
+        self.memory_outputs = memory_outputs
+
+    def forward(
+        self,
+        domain_pred_s,
+        domain_pred_t,
+        features_s,
+        features_t,
+        sample_idx_t,
+        **kwargs,
+    ):
+        """Compute the domain adaptation loss"""
+        domain_label = torch.zeros(
+            (domain_pred_s.size()[0]),
+            device=domain_pred_s.device,
+        )
+        domain_label_target = torch.ones(
+            (domain_pred_t.size()[0]),
+            device=domain_pred_t.device,
+        )
+
+        # update classification function
+        loss_adv = self.domain_criterion_(
+            domain_pred_s, domain_label
+        ) + self.domain_criterion_(domain_pred_t, domain_label_target)
```

Comment:
in eq (1) of the paper, they use both the cross entropy and the modified cross entropy loss (log(1-prob))

Reply:
we should also use label smoothing as the code of the paper: https://github.com/CrownX/SPA/blob/main/code/loss.py#L65-L84